### PR TITLE
[FLINK-26308][FLINK-26471] Separate Flink job role from operator role

### PR DESCRIPTION
**Pull request by mbalassi**
This separates the Flink job role from the operator role, which fixes two issues:
1. That when the `operatorNamespace` is the same as one of the `watchNamespaces` the install fails due to attempted duplicated role creation which I introduced in #35.
2. That deleting the operator role affected jobs described in FLINK-26471.

@wangyang0918 has raised another issue with #35 [here](https://github.com/apache/flink-kubernetes-operator/pull/35#issuecomment-1058750206), that prior to 1.15 when using `watchNamespaces` unless the user explicitly sets 
```
kubernetes.rest-service.exposed.type: ClusterIP
```
I would like to avoid overwriting the default Flink behavior in the operator, my suggestion is to present a warning to the user in the admission control validation and we might a line to the Readme / docs anyway.

Also while testing this change I have discovered that the webhook fails when we have multiple entries in `watchNamespaces`. I will fix these two issues in a separate PR early next week.

### CODE REVIEW

File: helm/flink-operator/values.yaml [hunk 1]
```diff
@@ -36,11 +36,16 @@ rbac:
...
 
-serviceAccount:
+operatorServiceAccount:
   create: true
   annotations: {}
   name: "flink-operator"
 
+jobServiceAccount:
+  create: true
+  annotations: {"helm.sh/resource-policy": keep}
```
Comment (gyfora) [inline helm/flink-operator/values.yaml:46]: could this simply be?
```
annotations: 
    "helm.sh/resource-policy" : keep
```

File: examples/basic.yaml [hunk 2]
```diff
@@ -19,14 +19,14 @@
 apiVersion: flink.apache.org/v1alpha1
 kind: FlinkDeployment
 metadata:
-  namespace: default
+  namespace: flink-user
```
Comment (gyfora) [inline examples/basic.yaml:22]: I think we should not change this in the example

**Review by gyfora — COMMENTED** [review]
Added some minor comments, I will test this manually tomorrow

File: helm/flink-operator/templates/rbac.yaml [hunk 3]
```diff
@@ -132,7 +190,25 @@ roleRef:
   apiGroup: rbac.authorization.k8s.io
 subjects:
   - kind: ServiceAccount
-    name: {{ template "flink-operator.serviceAccountName" . }}
+    name: {{ include "flink-operator.serviceAccountName" . }}
+    namespace: {{ .Values.operatorNamespace.name }}
+---
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (wangyang0918) [inline helm/flink-operator/templates/rbac.yaml:196]: Given that watchNamespaces is not configured, I am thinking whether we really need to create the `ClusterRole` and `ClusterRoleBinding` for flink job. Maybe `Role` and `RoleBinding` is enough. If users want to create a FlinkDeployment in a different namespace with flink-k8s-operator, then they should create the `serviceAccount` `Role` `RoleBinding` manually.

File: examples/basic-checkpoint-ha.yaml [hunk 4]
```diff
@@ -40,7 +40,7 @@ spec:
       cpu: 1
   podTemplate:
     spec:
-      serviceAccount: flink-operator
+      serviceAccount: flink
```
Comment (wangyang0918) [inline examples/basic-checkpoint-ha.yaml:43]: We could also update the serviceAccount in `e2e-tests/data/cr.yaml` to `flink`.

**Review by wangyang0918 — COMMENTED** [review]
Manually verified with/without `watchNamespaces` configured. It works well.
I just have left two comments.

File: helm/flink-operator/templates/rbac.yaml (hunk 3 above)
Comment (mbalassi) [reply helm/flink-operator/templates/rbac.yaml:196]: Fair point, given that we are trying to limit the footprint of the roles created I will create only role/rolebing in the operator namespace.

File: examples/basic-checkpoint-ha.yaml (hunk 4 above)
Comment (mbalassi) [reply examples/basic-checkpoint-ha.yaml:43]: Yes, great catch.

**Comment by mbalassi** [comment]
Thanks @wangyang0918 and @gyfora. I have addressed your comments.

**Comment by mbalassi** [comment]
I added an admittedly ugly fix for the webhook not working for multiple namespaces. @wangyang0918 maybe you have a nicer approach for this. The summary is that when we pass the `watchNamespaces` array through the template the output looks as follows:
```
namespaceSelector:
    matchExpressions:
      - key: kubernetes.io/metadata.name
        operator: In
        values: [flink-user default]
```
Which creates a selector that cannot be matched, the correct value would be `["flink-user", "default"]`.

**Pull request merged**